### ISSUE TITLE
Calling SetValue on Element/DateSelect with an empty string set's to todays date

### ISSUE THREAD

**Issue by michalbundyra** [laminas/laminas-form#54]
If a DateSelect element has it's value set to an empty string, due to the behaviour of php's date time object; the element is set to today's date. See offending code:

https://github.com/zendframework/zend-form/blob/29a930f8a66cad97b727be8bad65b1f2e1d288f8/src/Element/DateSelect.php#L106

This may be considered desirable; however if the option create_empty_option is set to create a default empty option, this behaviour prevents the element from being set back to the empty option. 

The fix would be to test for an empty string in that conditional, but is a potential BC break

---

Originally posted by @carnage at https://github.com/zendframework/zend-form/issues/8

**Comment by michalbundyra** [comment]
I don't think this is a bug, but rather expected behavior (in line with `DateTime`'s behavior).

I'd rather suggest accepting `null` and then nulling all fields if that is the case.


---

Originally posted by @Ocramius at https://github.com/zendframework/zend-form/issues/8#issuecomment-121618743

**Comment by fabiang** [comment]
> I'd rather suggest accepting `null` and then nulling all fields if that is the case.

Null is already possible and will result in the fields being empty. Unfortunately you get three warnings then:

```
Notice: Trying to access array offset on value of type null in /var/www/html/vendor/laminas/laminas-form/src/Element/DateSelect.php on line 121
Notice: Trying to access array offset on value of type null in /var/www/html/vendor/laminas/laminas-form/src/Element/DateSelect.php on line 122
Notice: Trying to access array offset on value of type null in /var/www/html/vendor/laminas/laminas-form/src/Element/DateSelect.php on line 123
```

Should we at least check for `$value === null` there?

**Comment by jroedel** [comment]
Yes! Please, check for nulls. This hasn't been plaguing my website for years! Would it help if I open a new issue?

**Comment by boesing** [comment]
> Yes! Please, check for nulls. This hasn't been plaguing my website for years! Would it help if I open a new issue?

Nope. It would probably help if you open a PR with providing a unit test. 👍🏻

**Comment by jroedel** [comment]
#70 

I added 2 tests clarifying the semantics of a null value. A null would be transformed into `today` (consistent with `DateTime` behavior) unless `create_empty_option` is `true` (in which the user has declared interest in nulls being an acceptable value).

**Comment by Slamdunk** [comment]
Closed by #118
